common/shamir-secret.c

```diff
@@ -0,0 +1,267 @@
+        mpz_xor(r, r, tmp);
+    }
+}
+
+/* Inverse of a number in 2^256 GF using Extended GCD */
+void inv_gf2256(mpz_t res, mpz_t a) {
+
+    if (mpz_cmp_ui(a, 0) == 0) {
+        return;
+    }
+
+    mpz_t r0, r1, s0, s1, q, temp1, temp2, tmp;
+    mpz_inits(r0, s0, s1, q, temp1, temp2, tmp, NULL);
+    mpz_t primitive_poly;
+    mpz_init_set_str(primitive_poly, prime_polynomial_str, 10);
+    mpz_set(r0, a);
+    mpz_init_set(r1, primitive_poly);
+    mpz_set_ui(s0, 1);
+    mpz_set_ui(s1, 0);
+
+    while(mpz_cmp_ui(r1, 0) > 0) {
+        div_gf2(q, r0, r1);
+        mpz_set(temp1, r1);
+        mul_gf2(tmp, q, r1);
+        mpz_xor(r1, r0, tmp);
+        mpz_set(r0, temp1);
+        mpz_set(temp2, s1);
+        mul_gf2(tmp, q, s1);
+        mpz_xor(s1, s0, tmp);
+        mpz_set(s0, temp2);        
+    }
+    mpz_set(res, s0);
+}
+
+/* Multiplication in the 2^256 Finite Field using BitMasking */
+void mul_gf2256(mpz_t result, mpz_t p, mpz_t q) {
+    mpz_t f1, f2, mask1, mask2, mask3, v, z, mod;
+    mpz_init_set_str(mod, prime_polynomial_str, 10);
+
+    mpz_inits(mask1, mask2, mask3, v, z, NULL);
+    mpz_init_set(f1, p);
+    mpz_init_set(f2, q);
+
+    if (mpz_cmp(f2, f1) > 0) {
+        mpz_swap(f1, f2);
+    }
+
+    if (mpz_cmp(mod, f1) == 0 || mpz_cmp(mod, f2) == 0) {
+        mpz_set_ui(result, 0);
+        return;
+    }
+
+    mpz_set_ui(mask1, 1);
+    mpz_mul_2exp(mask1, mask1, 256);
+    mpz_set(v, f1);
+    mpz_set_ui(z, 0);
+    while (mpz_cmp_ui(f2, 0) != 0) {
+        mpz_set_ui(mask2, mpz_tstbit(f2, 0));
+        if (mpz_cmp_ui(mask2, 0) > 0) {
+            char *msk = (char *)malloc(256 + 1);
+            memset(msk, '1', 256);
+            msk[256] = '\0';
+            mpz_set_str(mask2, msk, 2);
+            free(msk);
+        }
+
+        mpz_t tmp, tmp2;
+        mpz_inits(tmp, tmp2, NULL);
+        mpz_xor(tmp, z, v);
+        mpz_and(tmp, tmp, mask2);
+
+        mpz_sub(tmp2, mask1, mask2);
+        mpz_sub_ui(tmp2, tmp2, 1);
+        mpz_and(tmp2, z, tmp2);
+        mpz_ior(z, tmp, tmp2);
+
+        mpz_mul_2exp(v, v, 1);
+
+        mpz_div_2exp(tmp2, v, 256);
+        mpz_set_ui(mask3, mpz_tstbit(tmp2, 0));
+        if (mpz_cmp_ui(mask3, 0) > 0) {
+            char *msk = (char *)malloc(256 + 1);
+            memset(msk, '1', 256);
+            msk[256] = '\0';
+            mpz_set_str(mask3, msk, 2);
+            free(msk);
+        }
+
+        mpz_xor(tmp, v, mod);
+        mpz_and(tmp, tmp, mask3);
+
+        mpz_sub(tmp2, mask1, mask3);
+        mpz_sub_ui(tmp2, tmp2, 1);
+        mpz_and(tmp2, tmp2, v);
+        mpz_ior(tmp2, tmp, tmp2);
+        mpz_set(v, tmp2);
+
+        mpz_clear(tmp);
+        mpz_clear(tmp2);
+        mpz_div_2exp(f2, f2, 1);
+    }
+    mpz_set(result, z);
+    
+    mpz_clear(f1);
+    mpz_clear(f2);
+    mpz_clear(mask1);
+    mpz_clear(mask2);
+    mpz_clear(mask3);
+    mpz_clear(v);
+    mpz_clear(z);
+}
+
+
+struct point *split(int n, int t, char *secret) {
```

Comment:
I think caller should turn char * into bytes, so this should take a "const struct secret *secret".